drivers: bluetooth: hci: Introduce ST SPI protocol V2

HoZHel opened the pull request:
Update structure of spi.c to have a better and cleaner separation
between STMicroelectronics and Zephyr SPI protocol.

Update existing boards according to the new DTS approach.

Remove BT_SPI_BLUENRG as it is become redundant according to
the new changes.

Introduce STMicroelectronics SPI protocol V2 which is used in BlueNRG-1
and successor devices.

Add "st,hci-spi-v2.yaml" and use compatible property to enable
the protocol for the required boards.

Add BLE support for SensorTile.box board.

The new changes have been tested on 96b_carbon, disco_l475_iot1, b_l4s5i_iot01a,
stm32l562e_dk, x_nucleo_idb05a1, and sensortile_box boards.

avisconti commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -186,42 +182,88 @@ static void release_cs(void)
+#define READ_CONDITION false
+#define WRITE_DATA_CONDITION(ret, rx_first) (rx_first != 0U || ret)
+
+static int bt_spi_get_header(uint8_t op, uint8_t *size)
```
have you tested also the "else" side of the code?

HoZHel replied:
Yes. It has been tested by 96b_carbon board which uses STM32F401RET and nRF51822 as main MCU and BLE module respectively.

avisconti commented on boards/arm/sensortile_box/Kconfig.defconfig:
```diff
@@ -8,11 +8,27 @@ if BOARD_SENSORTILE_BOX
 	default "sensortile_box"
 
-if SPI
+if BT
```
Why not ```depends on BT``` here?

erwango replied:
`if` is a better option when you have several symbols impacted, as this is the case here.
`depends on` is well suited for single symbol dependency.
So current usage looks good to me.

avisconti commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -178,11 +178,12 @@ static void assert_cs(void)
-static void release_cs(void)
+static void release_cs(bool data_transaction)
 {
+	ARG_UNUSED(data_transaction);
```
not sure I got why you added an arg and declaring it unused. Is it for future?

avisconti replied:
ok, got it, sorry. It is because in v2 it is required...

avisconti commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -223,14 +224,89 @@ static int bt_spi_get_header(uint8_t op, uint8_t *size)
 	return ret;
 }
 
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_hci_spi_v2)
```
I wonder whether would be more readable to have all this routines as callbacks declared in separate files one for each protocol... Not sure it is worth doing it...

msmttchr replied:
@avisconti, let me share my point of view about this.
Protocol have all the same structure, that is:
- header phase: used for command and data size definition
- data phase: where data is exchanged

What is more different is that header phase, while data phase is similar.
So, it could be a good compromise to keep everything together.
On the other hand, @JordanYates raised your point in the past as well.

Not a strong opinion from me, but I am little bit more inclined to keep the single file.

avisconti replied:
@msmttchr yes, sure. If changes are not that much, it might be useless to separate into multiple files.

avisconti commented:
It seems a very good an well written PR to me. Want to take some time to review it deeply before approving it though.

JordanYates commented:
It would be nice to see a description of the differences between the two protocols (or a link to ST docs).
Does this affect two Zephyr devices using the built-in `samples/bluetooth/hci_spi` application?

msmttchr commented:
> It would be nice to see a description of the differences between the two protocols (or a link to ST docs). Does this affect two Zephyr devices using the built-in `samples/bluetooth/hci_spi` application?

Regarding the protocol description, you can have a look at:
- V1: https://www.st.com/resource/en/user_manual/um1865-the-bluenrgms-bluetooth-le-stack-application-command-interface-aci-stmicroelectronics.pdf (section 5)
- V2: https://www.st.com/resource/en/datasheet/bluenrg-2n.pdf (section 6.2)

Regarding the protocol implemented in `samples/bluetooth/hci_spi`, this is not impacted in term of behavior, but it is impacted in terms of code restructuring.

HoZHel committed:
dts: bindings: bluetooth: Add new yaml files for ST SPI protocol V1 and V2

Add "st,hci-spi-v1.yaml" to represent STMicroelectronics SPI protocol V1
which is used by BlueNRG-MS devices.

Add "st,hci-spi-v2.yaml" to show STMicroelectronics SPI protocol V2
utilized by BlueNRG-1 and successor devices.

Signed-off-by: Ali Hozhabri <ali.hozhabri@st.com>

HoZHel committed:
boards: arm: DTS update for ST boards that include BlueNRG-MS

Update DTS files of the STMicroelectronics boards which have BlueNRG-MS
as a BLE module.

Signed-off-by: Ali Hozhabri <ali.hozhabri@st.com>

HoZHel committed:
drivers: bluetooth: hci: Restructure spi.c to handle different protocols

Update structure of spi.c to have a better and cleaner separation
between STMicroelectronics and Zephyr SPI protocol.

Introduce bt_spi_get_header to separate algorithms well for
header acquisition in different protocols.

Remove header acquisition from bt_spi_rx_thread to make it simpler.

Signed-off-by: Ali Hozhabri <ali.hozhabri@st.com>

HoZHel committed:
boards: arm: sensortile_box: Add BLE support for SensorTile.box board

Add BLE support for SensorTile.box board.

Signed-off-by: Ali Hozhabri <ali.hozhabri@st.com>

HoZHel committed:
drivers: bluetooth: hci: Remove config BT_SPI_BLUENRG

Remove BT_SPI_BLUENRG Kconfig parameter as it is redundant according to
the new changes introduced by ST SPI protocol V1 and V2.

Remove "config BT_SPI_BLUENRG" from the boards that were using it.

Signed-off-by: Ali Hozhabri <ali.hozhabri@st.com>

HoZHel commented:
The changes have been split into more commits to make it easier for review.

HoZHel commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -167,16 +159,17 @@ static bool bt_spi_handle_vendor_evt(uint8_t *msg)
 	return handled;
 }
 
+#define IS_IRQ_HIGH gpio_pin_get_dt(&irq_gpio)
```
Yes, since their use cases are different, and also `READ_CONDITION` is set to false for two protocols.

HoZHel commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -223,14 +224,88 @@ static int bt_spi_get_header(uint8_t op, uint8_t *size)
+	uint64_t start = k_uptime_get();
+	uint64_t now = start;
+	/* Wait up to a maximum time of 100 ms */
+	while (!IS_IRQ_HIGH && (k_uptime_delta(&now) < 100)) {
```
I didn't know about it. Thanks for informing.

HoZHel replied:
It has been replaced by `WAIT_FOR`.

HoZHel committed:
drivers: bluetooth: hci: Remove header acquisition from bt_spi_send

Replace the header acquisition scope in bt_spi_send with bt_spi_get_header.

Add WRITE_DATA_CONDITION in order not to retransmit data without reading
the header as it is meaningless for ST BlueNRG devices.

Signed-off-by: Ali Hozhabri <ali.hozhabri@st.com>

HoZHel committed:
drivers: bluetooth: hci: Introduce ST SPI protocol V2

Introduce STMicroelectronics SPI protocol V2 which is used in BlueNRG-1
and successor devices.

Change the size of the variable "size" to 16 bits as it is necessary for ST
BlueNRG-1 and successor devices.

Signed-off-by: Ali Hozhabri <ali.hozhabri@st.com>

erwango reviewed (dismissed):
LGTM for the parts I'm involved.

jori-nordic commented:
@JordanYates you might want to review this, as it is a refactor of `spi.c`, not just a new driver.

erwango commented:
Current twister issue should be fixed by https://github.com/zephyrproject-rtos/zephyr/pull/65540

avisconti reviewed (dismissed):
LGTM

HoZHel commented:
@jhedberg Would you please review the PR?

carlescufi commented:
@JordanYates @jhedberg could you take a look please?

jhedberg commented:
Looks like this will conflict with #65179. What's the plan to sync these two PRs? @arbrauns FYI.

HoZHel commented:
> Looks like this will conflict with #65179. What's the plan to sync these two PRs? @arbrauns FYI.

Well, whichever is merged first, the other will rebase, resolve conflict, and force-push.

jhedberg commented:
Hmm, GitHub doesn't indicate any conflicts, so maybe a rebase isn't even necessary. Would be good to double-check that nothing is broken with the combination of the two PRs, though.

alwa-nordic commented on samples/bluetooth/hci_usb_h4/sample.yaml:
```diff
@@ -10,4 +10,4 @@ tests:
       - bluetooth
     # FIXME: exclude due to build error
-    platform_exclude: 96b_carbon stm32l562e_dk
+    platform_exclude: 96b_carbon stm32l562e_dk sensortile_box
```
You are adding onto a 'FIXME'. Why is this needed? Can it be resolved properly in some way?

HoZHel replied:
This test is related to Bluetooth over USB, and it is performed by Twister automatically when the board has "usb_device" and "ble" capability in its yaml file. However, this dependency for the test is not true that it indicates every board which has aforementioned capabilities should support Bluetooth over USB. At least, at the moment SensorTile.box does not support this feature. In addition, the comment "FIXME" is wrong as there is no problem to be fixed.

I may skip this test without using exclusion since there is already a misspelling in the board's yaml file regarding "usb_device" parameter, and intentionally or by mistake instead of "_" there is a white space.
Please, let me know which approach you prefer.

HoZHel commented:
> Hmm, GitHub doesn't indicate any conflicts, so maybe a rebase isn't even necessary. Would be good to double-check that nothing is broken with the combination of the two PRs, though.

Even though we changed the same file, we didn't touch the same parts. So, generally there shouldn't be any conflict; however, I'll rebase and force push to double-check.

JordanYates commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -411,37 +504,31 @@ static int bt_spi_send(struct net_buf *buf)
-	} while ((header_rx[STATUS_HEADER_READY] != READY_NOW ||
-		  (header_rx[1] | header_rx[2] | header_rx[3] | header_rx[4]) == 0U) && !ret);
+	if (size < buf->len) {
+		LOG_WRN("Unable to write full data");
```
Writing partial data is worse than writing no data at all IMO.
This check can also never fail for the non-ST protocol, not sure whether the compiler is smart enough to figure that out though.

HoZHel replied:
Done.

JordanYates reviewed (changes requested):
Minor problem with writing partial packets.

Not part of the requested changes, but this file now contains enough `#ifdef` layers that I find it annoying to figure out which parts apply when. I think it would be worthwhile trying to split the various categories into dedicated files soon.

erwango reviewed (dismissed):
Approved on the parts I'm involved

msmttchr reviewed (approved):
LGTM. Thanks.

msmttchr commented:
@jhedberg, unfortunately this PR is quite urgent for me, as it can unblock a certain number of boards to be supported in Zephyr.
Apparently, all the comments have been addressed, do you see any blocking point for merging it?

jhedberg commented:
> @jhedberg, unfortunately this PR is quite urgent for me, as it can unblock a certain number of boards to be supported in Zephyr. Apparently, all the comments have been addressed, do you see any blocking point for merging it?

No, I don't see any showstoppers for merging. Not sure why it hasn't happened yet, since typically people from the release engineering team (with merge privileges) periodically check for PRs that are mergeable and merge them. @erwango I see you requested @fabiobaltieri for review - is there something still unclear or that was just to expedite getting this merged (I also have merge privileges, but wanted to check before I press "merge").

jhedberg commented:
> No, I don't see any showstoppers for merging. Not sure why it hasn't happened yet, since typically people from the release engineering team (with merge privileges) periodically check for PRs that are mergeable and merge them. 

Answering myself:
"1 change requested

@JordanYates requested changes"

That's the reason the PR didn't show up in any "mergeable PR" filters.

erwango commented:
> @erwango I see you requested @fabiobaltieri for review - is there something still unclear or that was just to expedite getting this merged (I also have merge privileges, but wanted to check before I press "merge").

The reason I added Fabio is that I though that maybe you were prevented to merge as you also are the assignee.

fabiobaltieri commented:
> The reason I added Fabio is that I though that maybe you were prevented to merge as you also are the assignee.

Not a thing, the only limitation on individuals is something like: no author and reviewer and merger from the same company.

@JordanYates is indeed blocking though, can you make another pass?
